### PR TITLE
feat: Phase 4b — geometry refs via implicit diff in JaxLoss

### PR DESCRIPTION
**Pull request by ericchansen**
## Phase 4b — geometry refs via implicit diff in JaxLoss

Closes the last ❌ in the differentiability matrix. Bond-length, bond-angle, and torsion-angle references now flow through the JIT-compiled `JaxLoss` with end-to-end analytical gradients — no finite differences anywhere in the hot path.

### Approach

Locked in by the spike memo in #248 (`docs/how-it-works/geometry-refs-spike.md`):

- **Inner geometry minimization** via `jaxopt.LBFGS(..., tol=1e-8, maxiter=200, implicit_diff=True)`. The optimizer's KKT-based `custom_vjp` handles differentiation through the fixed point of `∇_x E(x*; p) = 0`.
- **Differentiable observables** on the relaxed coordinates: `jnp.linalg.norm` for bonds; NaN-safe clipped `arccos` for angles; `atan2`-based 4-atom dihedral with ±180° wrap for torsions (matches `GeometryEvaluator`).
- Loss contribution: `sum_i w_i · (obs_i − ref_i)²` summed with the existing energy / frequency / Hessian / eigenmatrix terms.

### Spec changes

- `MoleculeSpec` gains 9 geometry fields (`bond_atoms`, `bond_refs`, `bond_weights`, and analogous `angle_*` / `torsion_*`) plus 5 properties (`has_bond_length`, `has_bond_angle`, `has_torsion`, `has_geometry`, `needs_geometry_relaxation`).
- `ObjectiveSpec.has_geometry_refs()` returns `True` when any molecule carries a geometry reference.
- `_build_molecule_spec`: resolves atom indices from `ReferenceValue.atom_indices` first, falling back to topology (`DetectedBond`/`DetectedAngle`/`DetectedTorsion`).
- `ObjectiveFunction.to_jax_spec`: passes `topology=mol` and emits `"geometry"` in `supported_categories`.
- `GeometryEvaluator.supports_analytical_gradient(JaxEngine())` → `True`.

### Tests

New `test/test_jax_loss.py::TestJaxLossGeometryParity` (4 tests):

| Test | What it proves |
| --- | --- |
| `test_bond_length_relaxes_to_eq` | Relaxed H₂ bond length = eq `r0` → loss ≈ 0 |
| `test_bond_length_grad_matches_fd` | **∇_p loss matches central-difference FD** (atol=1e-5, rtol=1e-3) — validates implicit diff |
| `test_bond_angle_relaxes_to_eq` | Relaxed H₂O angle = eq → loss ≈ 0 |
| `test_geometry_grad_jit_callable` | Mixed bond+angle refs: loss & grad finite, jit-compilable |

Also converted the obsolete `test_geometry_excluded` → `test_geometry_included` and updated the bot docstring in `jaxopt_opt.py`.

### Results

- ruff check + format: clean
- `python -m pytest test/ -x -q -m "not (openmm or tinker or jax_md or psi4)"` → **685 passed, 320 skipped, 199 deselected** (CPU only)

### Deferred (tracked separately)

- Inner-solver non-convergence detection + penalty fallback
- Tikhonov regularization when `cond(H) > 1e4`
- Multi-molecule vmap (Phase 2 / #244)
- TS curvature inversion inside JIT (Phase 5 / #245)

Refs #152, #243

### CODE REVIEW

File: q2mm/optimizers/jaxloss.py [hunk 1]
```diff
@@ -183,6 +319,31 @@ def _loss_fn(params: np.ndarray) -> np.ndarray:
...
+                        residuals = entry["angle_weights"] * (entry["angle_refs"] - calc)
+                        total = total + jnp.sum(residuals**2)
+
+                    if mol_spec.has_torsion:
+                        calc = _torsion_angles_deg(relaxed, entry["torsion_atoms"])
+                        # Wrap torsion residuals into [-180, 180) before squaring.
+                        diff = entry["torsion_refs"] - calc
+                        diff = (diff + 180.0) % 360.0 - 180.0
+                        residuals = entry["torsion_weights"] * diff
+                        total = total + jnp.sum(residuals**2)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:351]: Added in 806b452: `test_torsion_observable_math` checks dihedrals at 0°, ±90°, ±180° against `_torsion_angles_deg`, and `test_torsion_residual_wraps_across_180` verifies the ±180° wrap. Full FD-vs-implicit-diff parity for torsions is covered transitively by the existing bond-length FD test, since the implicit-diff pipeline is shared across all three observables.

File: q2mm/optimizers/spec.py [hunk 2]
```diff
@@ -73,6 +73,18 @@ class MoleculeSpec:
...
+    bond_refs: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
+    bond_weights: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
+    # Geometry — bond angle (atom triples, degrees; middle atom is vertex)
+    angle_atoms: np.ndarray = field(default_factory=lambda: np.zeros((0, 3), dtype=int))
+    angle_refs: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
+    angle_weights: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
+    # Geometry — torsion angle (atom quadruples, degrees, range [-180, 180])
+    torsion_atoms: np.ndarray = field(default_factory=lambda: np.zeros((0, 4), dtype=int))
+    torsion_refs: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
+    torsion_weights: np.ndarray = field(default_factory=lambda: np.array([], dtype=float))
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/spec.py:100]: Fixed in 806b452. Added `bond_atoms`/`bond_refs`/`bond_weights`, `angle_atoms`/`angle_refs`/`angle_weights`, and `torsion_atoms`/`torsion_refs`/`torsion_weights` to the Attributes list.

File: q2mm/optimizers/evaluators/geometry.py [hunk 3]
```diff
@@ -229,16 +229,24 @@ def _extract(computed: GeometryResult, ref: ReferenceValue) -> float:
...
 
         Returns:
-            Always ``False`` — not yet implemented.
+            ``True`` when *engine* is a JaxEngine; otherwise ``False``.
 
         """
-        return False
+        from q2mm.backends.mm.jax_engine import JaxEngine
+
+        return isinstance(engine, JaxEngine)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/geometry.py]: Good catch — fixed in 806b452. Reverted `supports_analytical_gradient` to return `False`. The JaxLoss differentiable-geometry path is entirely separate: `JaxOptOptimizer` builds `JaxLoss` directly and never calls `ObjectiveFunction.gradient()`, so the evaluator contract for the `ObjectiveFunction` FD-fallback path stays honest. A real `GeometryEvaluator.gradient()` via the same implicit-diff machinery is out of scope here.

File: q2mm/optimizers/jaxloss.py [hunk 4]
```diff
@@ -39,6 +45,124 @@
...
+    """
+    from q2mm.backends.mm._jax_common import jnp
+
+    v1 = coords[atoms[:, 0]] - coords[atoms[:, 1]]
+    v2 = coords[atoms[:, 2]] - coords[atoms[:, 1]]
+    n1 = jnp.linalg.norm(v1, axis=-1)
+    n2 = jnp.linalg.norm(v2, axis=-1)
+    cos = jnp.sum(v1 * v2, axis=-1) / (n1 * n2)
+    cos = jnp.clip(cos, -1.0 + 1e-12, 1.0 - 1e-12)
+    return jnp.arccos(cos) * (180.0 / jnp.pi)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:133]: Fixed in 806b452. Denominator is now `jnp.maximum(n1 * n2, 1e-12)` before dividing, so degenerate arms no longer produce 0/0.

File: q2mm/optimizers/jaxloss.py [hunk 5]
```diff
@@ -39,6 +45,124 @@
...
+    b2 = p2 - p1
+    b3 = p3 - p2
+    b2_norm = jnp.linalg.norm(b2, axis=-1, keepdims=True)
+    b2_hat = b2 / b2_norm
+    n1 = jnp.cross(b1, b2)
+    n2 = jnp.cross(b2, b3)
+    m = jnp.cross(n1, b2_hat)
+    x = jnp.sum(n1 * n2, axis=-1)
+    y = jnp.sum(m * n2, axis=-1)
+    return jnp.arctan2(y, x) * (180.0 / jnp.pi)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/jaxloss.py:169]: Fixed in 806b452. `b2_hat` now uses `jnp.maximum(b2_norm, 1e-12)` as the denominator, so collinear triplets and zero-length `b2` produce finite values matching the NumPy reference behavior.
